Add Galera WSREP Provider

tomalok opened the pull request:
* download, patch, build, install, and strip Galera WSREP Provider
* patches allow Galera to build on Alpine/musl for x86/x86_64 architectures (patches would need tweaks to support other arch's)
* adds NPROC make variable (my builds were unstable with `-j $(nproc)`)
* set default ALPINE_VER to 3.11

tomalok committed:
Add Galera WSREP Provider

* download, patch, build, install, and strip Galera WSREP Provider
* patches allow Galera to build on Alpine/musl for x86/x86_64 architectures (patches would need tweaks to support other arch's)
* adds NPROC make variable (my builds were unstable with `-j $(nproc)`)
* set default ALPINE_VER to 3.11

csandanov commented:
Thanks. A couple things:
- how much does it increase the compressed image size? I guess not much but want to know beforehand (we could use a separate tag for Galera)
- I think alpine upgrade should be a separate update, we still have 3.8 for 10.1 and 3.10 for the rest ( see `./circleci`). Or is it important to have 3.11 for galera?

tomalok commented:
* The only additional artifact in the image is `/usr/lib/libgalera_smm.so`, which is 2,634,224 bytes, stripped.  The Galera build process also produces a `garbd` (galera arbitrator) daemon to provide quorum for clusters with even number of nodes (and a man page for that daemon)...  For Alpine, I bundled those in separate packages, as they have no dependencies on the library and didn't make sense to put in the `wodby/mariadb` image.
* There is no requirement for Alpine 3.11, but personally, I like to keep as current as possible. I can roll back ALPINE_VER back in this PR if you like.

csandanov commented:
Yes, could you please rollback the Alpine version, we will perform the update separately. 

Also, it's not really clear how to use galera with this image, if you look up helm chart for galera, they have a specific binary that runs in the entrypoint and searching for other galera master instances in the network and connects them all together (is it garbd?). If we want people to run galera with this image we should document it in README.

tomalok commented:
Okay, things for to do, off the top of my head yet...

- [x] Roll back the Alpine version
- [x] Make build of Galera optional
- [ ] ~Per [Galera Versions](https://mariadb.com/kb/en/what-is-mariadb-galera-cluster/#galera-versions) builds of 10.1, 10.2, and 10.3 should use Galera 25.3.28 (10.5.1 should use 26.4.4, but you're not tracking 10.5 yet)~
- [x] Maybe include `garbd`?  A search on https://hub.docker.com didn't turn up many hits, and only one is recent.
- [x] Add `WSREP_` configuration to the template
- [x] Add some docs about how to start a cluster, and perhaps about `garbd`

I'm only seeing `linux/amd64` OS/arch values for `wodby/mariadb`, so it should not be a blocker that the musl patches don't work (yet) for non-x86 arch.

tomalok committed:
* ALPINE_VER back to 3.10
* make building Galera optional
* also install /usr/sbin/garbd
* add a job to .circleci/config.yml to create a 10.4-galera image

tomalok commented:
Ran into some complications building Galera with MariaDB versions prior to 10.4...  Dropping that item off the to-do list.

tomalok committed:
* Galera SST methods need rsync and/or socat
* Update 10.4 my.cnf.tmpl for Galera configs
* Update README.md with Galera env vars

tomalok commented:
* needed `rsync` and `socat` for Galera SST.
* updated 10.4 my.cnf.tmpl to include Galera `WSREP_*` configs and enforce some other configs when `WSREP_ON` env var exists.  (Is there a way with `gotpl` to specify that it exists and its value is `ON`?)
* added `WSREP_*` vars to `README.md`

tomalok committed:
* `garbd` requires boost-program_options package
* update README - galera image tags & galera cluster documentation

tomalok commented:
The latest commit adds a runtime dependency for `garbd`, and fleshes out the rest of the Galera-related documentation in the README.

@csandanov - unless there's anything else you think needs tweaking, I think this PR is ready.

csandanov commented:
Thank you, looks good. We'll have to add tests with wsrep enabled at some point.
